src/compute-client/src/service.rs

```diff
@@ -85,16 +85,53 @@ where
 
 /// Maintained state for partitioned compute clients.
 ///
-/// This helper type unifies the responses of multiple partitioned
-/// workers in order to present as a single worker.
+/// This helper type unifies the responses of multiple partitioned workers in order to present as a
+/// single worker:
+///
+///   * It emits `FrontierUppers` responses reporting the minimum/meet of frontiers reported by the
+///     individual workers.
+///   * It emits `PeekResponse`s and `SubscribeResponse`s reporting the union of the responses
+///     received from the workers.
+///
+/// In the compute communication stack, this client is instantiated several times:
+///
+///   * One instance on the controller side, dispatching between cluster processes.
+///   * One instance in each cluster process, dispatching between timely worker threads.
+///
+/// Note that because compute commands, except `CreateTimely`, are only sent to the first process,
+/// the cluster-side instances of `PartitionedComputeState` are not guaranteed to see all compute
+/// commands. Or more specifically: The instance running inside process 0 sees all commands,
+/// whereas the instances running inside the other processes only see `CreateTimely`. The
+/// `PartitionedComputeState` implementation must be able to cope with this limited visiblity. It
+/// does so by performing most of its state management based on observed compute responses rather
+/// than commands.
 #[derive(Debug)]
 pub struct PartitionedComputeState<T> {
     /// Number of partitions the state machine represents.
     parts: usize,
     /// Upper frontiers for indexes and sinks, both unioned across all partitions and from each
     /// individual partition.
+    ///
```

Comment:
Why does the above say that the frontiers are "unioned across all partitions". I was expecting to see the meet or the join of all frontiers... but what is the definition and the interpretation of the "union" of a collection of frontiers?

Reply:
Yeah, it's the meet. Basically, all frontiers are "unioned" by adding them to the `MutableAntichain` and then calling `.frontier()` on that, which gives you the greatest lower bound of all contained frontiers. I'm not sure if "union" of frontiers makes much sense outside the context of `MutableAntichain`.

Reply:
Calling it the meet wouldn't be accurate though, because we are not just storing the meet, we store the `MutableAntichain`. That is more than the meet, as it stores and provides access to the individual frontiers too.

Reply:
Perhaps one could write here: "both collected as an antichain across all partitions and individually listed for each partition."

Reply:
Hm, the more I think about it, the more "unioned frontiers" grows on me. That a `MutableAntichain` is used for the union should be immediately clear when looking at the type, so I don't see a need to spell it out again. Unless someone feels strongly about this, in which case I'd be happy to do the spelling out!

Reply:
What I feel strongly about is that when I read "unioned frontiers" I imagine something like:

$$\bigcup_i \mathit{F}_i$$

where

$$\\{ \mathit{F}_i \\}_{i \in I}$$

is some indexed set of frontiers. Since each frontier is an antichain (that is, a set of mutually incomparable elements), taking the union of all these antichains is not guaranteed to be an antichain.

You need to do some cleanup to remove redundant elements, as evidenced [by the definition of that operation](https://en.wikipedia.org/wiki/Antichain#Join_and_meet_operations). It makes sense to talk about the "union of the lower sets corresponding to a family of antichains", but not about the "union of antichains" - the corresponding operation is the meet.

Maybe just by looking at the name of the corresponding type (`MutableAntichain`) one would be able to infer that the union won't contain $x$ and $y$ such that $x \leq y$, but for me this is additional mental burden caused by imprecise language in the documentation.

Reply:
FYI I'm fine with this as well (as the notion of a `union` is somehow well-defined on a `MutableAntichain`, and extracting the actual antichain is done by calling `frontier()`):
> `MutableAntichain`s used to maintain the upper frontiers for indexes and sinks

Reply:
I think the disconnect here is that internally, a `MutableAntichain` *is* in fact a union of (the elements) of the individual frontiers, not just their antichain. It needs to retain the individual frontiers to enable advancing and retracting them individually.

Externally, a `MutableAntichain` presents as an antichain, hence the name I assume. But if we called it an "antichain" in the comments, I would be worried for people to confuse that with the `Antichain` type. But we don't store an `Antichain` (or meet), we store more than that.

I think you have the mental burden of needing to understand what `MutableAntichain` does either way. But it seems that you and Marcos are actively confused about the word "unioned" because it makes you think of set theory. Can we agree to use another general term instead? Marcos suggested "collected". Alternatively we could spell out `MutableAntichain`, which is redundant with the type definition, but better that than confusing.

Reply:
> Alternatively we could spell out `MutableAntichain`, which is redundant with the type definition, but better that than confusing.

I agree with that (see my last comment)!

Reply:
> Out of curiosity: is the problem that saying

I'm not worried about things being to mathematical. Rather that storing a meet / lower bound of frontiers would mean storing an `Antichain`, which is not what we are doing.

Reply:
Thanks, I got your point.

Reply:
Changed!